gtda/utils/validation.py

```diff
@@ -204,45 +220,90 @@ def check_point_clouds(X, distance_matrix=False, **kwargs):
+            raise ValueError(
+                f"Input array X must have X.shape[1] == X.shape[2]: "
+                f"{X.shape[1]} != {X.shape[2]} passed.")
+        elif (X.shape[1] == X.shape[2]) and not distance_matrices:
+            warn(
+                "Input array X has X.shape[1] == X.shape[2]. This is "
+                "consistent with a collection of distance/adjacency "
+                "matrices, but the input is being treated as a collection "
+                "of vectors in Euclidean space.",
+                DataDimensionalityWarning, stacklevel=2)
+        Xnew = _check_array_mod(X, **kwargs_, allow_nd=True)
     else:
-        if not distance_matrix:
-            reference = X[0].shape[1]  # Embedding dimension of first sample
-            if not reduce(
-                    and_, (x.shape[1] == reference for x in X[1:]), True):
-                warn("Not all point clouds have the same embedding dimension.")
-
-    has_check_failed = False
-    messages = []
-    Xnew = []
-    for i, x in enumerate(X):
-        try:
-            Xnew.append(check_array(x, **kwargs_))
-            messages = ['']
-        except ValueError as e:
-            has_check_failed = True
-            messages.append(str(e))
-    if has_check_failed:
-        raise ValueError("The following errors were raised by the inputs: \n"
-                         "\n".join(messages))
+        has_check_failed = False
+        messages = []
+        Xnew = []
+        for i, x in enumerate(X):
+            try:
+                xnew = _check_array_mod(x, **kwargs_, ensure_2d=True)
+                if distance_matrices:
+                    if not x.shape[0] == x.shape[1]:
+                        raise ValueError(
+                            f"All arrays must be square: {x.shape[0]} rows "
+                            f"and {x.shape[1]} columns found in this array.")
+                Xnew.append(xnew)
+            except ValueError as e:
+                has_check_failed = True
+                messages.append(f"Entry {i}:\n{e}")
+        if has_check_failed:
+            raise ValueError(
+                "The following errors were raised by the inputs:\n\n" +
+                "\n\n".join(messages))
+
+        if not distance_matrices:
+            if reduce(and_, (x.shape[0] == x.shape[1] for x in X), True):
+                warn(
+                    "All arrays are square. This is consistent with a "
+                    "collection of distance/adjacency matrices, but the input "
+                    "is being treated as a collection of vectors in Euclidean "
+                    "space.", DataDimensionalityWarning, stacklevel=2)
+
+            ref_dim = X[0].shape[1]  # Embedding dimension of first sample
+            if not reduce(and_, (x.shape[1] == ref_dim for x in X[1:]), True):
+                warn(
+                    "Not all point clouds have the same embedding dimension.",
+                    DataDimensionalityWarning, stacklevel=2)
```

Comment:
Could you detail the choice of `stacklevel` or point me to a reference, please? I am completely unfamiliar with that and could not find anything.

Reply:
I basically stole this from sklearn's `check_array`, see e.g. https://github.com/scikit-learn/scikit-learn/blob/95d4f0841d57e8b5f6b2a570312e9d832e69debc/sklearn/utils/validation.py#L450.